### PR TITLE
Annotate `ServerConfig` to fix Gradle deprecation warnings

### PR DESCRIPTION
**Pull request by wolfs**
This is another part of https://github.com/gretty-gradle-plugin/gretty/issues/133.

See the discussion on https://github.com/gretty-gradle-plugin/gretty/pull/194 as well.

### CODE REVIEW

File: libs/gretty/build.gradle [hunk 1]
```diff
@@ -1,4 +1,5 @@
 plugins {
+  id 'java-gradle-plugin'
```
Comment (f4lco) [inline libs/gretty/build.gradle:2]: I can't really see the implications of the plugin - they might be manifold. Is the desired effect the same as 80d8dddc5a0fa5744e77755044043faad0bf316c? Or setting `org.gradle.warning.mode=fail` for that matter?

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+
+import java.util.List;
+import java.util.Map;
+
+public interface ServerConfigWithInputs {
```
Comment (f4lco) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:16]: Please don't mix. Either

1. place .java source files in src/main/java, or
2. make this file .groovy and apply some groovyfication (stripping `;`, for example)

I would suggest (2), it's a mostly Groovy project, and a groovy-only module.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    @Input
+    Integer getHttpsIdleTimeout();
+
+    @Optional
+    @Input
+    String getSslHost();
+
+    @Optional
+    @InputFile
+    @PathSensitive(PathSensitivity.NONE)
```
Comment (f4lco) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:67]: Why do you think is `PathSensitivity.NONE` appropriate? Also for the properties of this interface.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    Object getSslKeyStorePath();
+
+    @Internal
+    String getSslKeyStorePassword();
+
+    @Internal
```
Comment (f4lco) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:73]: This seems to be the way to ignore a property. Nice, I wish this was mentioned in the docs in a more prominent place. I couldn't find it.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    @Input
+    String getSslHost();
+
+    @Optional
+    @InputFile
+    @PathSensitive(PathSensitivity.NONE)
+    Object getSslKeyStorePath();
+
+    @Internal
+    String getSslKeyStorePassword();
```
Comment (f4lco) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:71]: Is this a security concern? If that password is not part of the inputs and then changed, the task might not re-execute, is that correct?

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    boolean getSslNeedClientAuth();
+
+    @Internal
+    Object getRealm();
+
+    @Internal
+    Object getRealmConfigFile();
+
+    @Internal
+    Object getServerConfigFile();
```
Comment (f4lco) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:94]: Why are realm, realmConfigFile, serverConfigFile not considered input?

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+    @Optional @Input
+    Integer getScanInterval();
+
+    @Optional
+    @InputFile
+    @PathSensitive(PathSensitivity.NONE)
+    Object getLogbackConfigFile();
+
+    @Console
```
Comment (f4lco) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:107]: That's cool, I did not know Gradle made that distinction.

**Review by f4lco — CHANGES_REQUESTED** [review]
Thank you so much doing this 💯  
You worked out a good approach. Most comments are minor or do not require fixing. Most of my concern goes towards the new plugin application. Without it, the commit might be more backport-friendly towards 3.x.

File: libs/gretty/build.gradle (hunk 1 above)
Comment (wolfs) [reply libs/gretty/build.gradle]: I removed the plugin application. I'll do that in a separate PR.

In short, this plugin sets up the build for building a Gradle plugin. So it configures publishing tasks and test tasks and the like. More importantly for this PR, it creates a `validatePlugins` task which checks that all tasks are annotated correctly.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 3 above)
Comment (wolfs) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:67]: It seems like the name of the logback config file is not important, only its contents. WDYT?

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 4 above)
Comment (wolfs) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:73]: Yeah, it is somewhat hidden in this table: https://docs.gradle.org/current/userguide/more_about_tasks.html#table:incremental_build_annotations

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 5 above)
Comment (wolfs) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:71]: Passwords should never be inputs. The output of the task does not depend on the password. Only whether the task fails or not depends on the password. And yes, this is also a security concern.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 6 above)
Comment (wolfs) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:94]: They are objects and I didn't know what types they are. I am happy to change the annotation, if you can give me some guidance what the annotations should be.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 7 above)
Comment (wolfs) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:107]: The annotation is only for documentation.

**Comment by wolfs** [comment]
@f4lco Thank you for the review! Maybe a general comment on the effect of the input/output annotations:
None of the tasks has an output file, so they will always be executed no matter if the inputs changed or not. A user of the plugin could add an output file to one of the tasks and enable up-to-date checks for them. Then the input annotations would matter.

We could decide that annotating all the properties with `@Internal` is also fine, though that would make the live for folks who want to make the tasks work with up-to-date checks harder.

File: libs/gretty/build.gradle (hunk 1 above)
Comment (f4lco) [reply libs/gretty/build.gradle]: > So it configures publishing tasks and test tasks and the like.

Yes, I suspected that. In theory we have all of that, the patch replace / beef up the existing packaging needs to take that into account. I think it needs a little more coordination. Also see #192, @javabrett wants to touch publishing anyway.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    @Optional @Input
+    String getScanner();
+
+    @Optional @Input
+    String getPortPropertiesFileName();
+
+    @Optional @Input
+    Boolean getLiveReloadEnabled();
+
+    @ReplacedBy("serverConfigFile")
```
[the remark this answers is not in the document]
Comment (boris-petrov) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:167]: Yes, that's a good idea!

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 5 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java:71]: That's interesting, never thought of it that way. When thinking more, this also cannot fail: likely changing the password also changes the file (due to encryption with the new password), which is again tracked by Gradle, right?

**Comment by boris-petrov** [comment]
Are we talking only about the `farmStart` and `appStart` tasks? Why would anyone need them to be up-to-date and not execute? I don't think we need this so `@Internal` should be fine in that case.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 2 above)
Comment (wolfs) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java]: I did 2)

**Comment by wolfs** [comment]
@f4lco I removed everything except the `ServerConfig` changes. Please take another look. I'll create a PR with the remaining fixes when this PR has been merged.

**Comment by boris-petrov** [comment]
Shouldn't `AppStartTask` also implement the interface?

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 5 above)
Comment (wolfs) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java]: Yes.

**Comment by wolfs** [comment]
@f4lco @boris-petrov Anything else I should address to get this merged?

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 3 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java]: That's something I still need to wrap my head around. Thanks for the explanation, for me it's a long stretch for `PathSensitivity.NONE` to "track file by content, not by name".  Your judgement seems reasonable!

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java (hunk 6 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java]: My guide post is `JettyConfigurer`:

```
void configureSecurity(context, String realm, String realmConfigFile, boolean singleSignOn)
```

and for `serverConfigFile` - for  whatever reason - does not seem to contrained, so `Object`.

Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/ServerConfigWithInputs.java]: In the light of the explanation of up-to-date checks below it does seem less relevant. LGTM.

**Review by f4lco — APPROVED** [review]

**Comment by f4lco** [comment]
All of my concerns have been addressed. @boris-petrov please see for yourself and merge at will. Thanks again @wolfs 😃

**Comment by boris-petrov** [comment]
Squashed and merged, thank you both!

**Pull request merged**